### PR TITLE
Master+organizations users+roles permissions

### PR DESCRIPTION
**Pull request by ddurieux**
Changes proposed in this pull request:

- Add roles management
- assign roles to users
- manage permissions on data (type of items and properties)
- manage permissions on structure (types and properties)
- update order of fill data when install (in folder `db/migrations/`)

How to test the feature manually:

1. create a role
2. adjust the permissions
3. assign the role to this user
4. get token for this user
5. do request on endpoints not have access and see 401 return code with the error message

Pull request checklist:

- [x] code is manually tested
- [x] tests are updated
- [x] commit messages are relevant
- [x] documentation is updated (including code comments, commit messages…)

_If you think one of the item isn’t applicable to the PR, please check it
anyway and precise `N/A` next to it._

Related to #

### CODE REVIEW

File: src/v1/Models/Item.php [hunk 1]
```diff
@@ -218,6 +218,15 @@ public function propertygroups()
     // return $this->belongsToMany('App\v1\Models\Config\Propertygroup', 'id', 'type_id');
   }
 
+  public function roles()
```
Comment (marien-probesys) [inline src/v1/Models/Item.php:221]: We know that an item can be attached to only one role (thanks to the unique index on the couple `role_id, item_id` on the `item_role` table). So I would rather rename `roles()` to `role()` and return only one role.

File: src/v1/Controllers/Token.php [hunk 2]
```diff
@@ -165,6 +165,17 @@ public function postToken(Request $request, Response $response, $args): Response
     $future = new DateTime("+20 minutes");
     // For test / DEBUG
     // $future = new DateTime("+30 seconds");
+    // Get roles
+    $roleId = null;
+    foreach ($user->roles()->get() as $role)
+    {
+      $roleId = $role->id;
+      break;
+    }
```
Comment (marien-probesys) [inline src/v1/Controllers/Token.php]: You really like loops :p

```suggestion
    $role = $user->roles()->first();
```

Or simply (if you go with my suggestion to rename `roles()` to `role()`):

```suggestion
    $role = $user->role();
```

Then check if `$role` is null line 175 and call `$role->id` line 187.

File: src/v1/Models/Item.php [hunk 3]
```diff
@@ -218,6 +218,15 @@ public function propertygroups()
...
+  {
+    if ($this->attributes['type_id'] != 2)
```
Comment (marien-probesys) [inline src/v1/Models/Item.php]: I think we should have global constants for these kind of ids (e.g. `TYPE_ORGANIZATION_ID` and `TYPE_USER_ID`)

File: src/v1/Models/Config/Role.php [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+  public function getUpdatedByAttribute($value)
+  {
+    return \App\v1\Models\Common::getUserAttributes($value);
+  }
+
+  public function getDeletedByAttribute($value)
+  {
+    return \App\v1\Models\Common::getUserAttributes($value);
+  }
```
Comment (marien-probesys) [inline src/v1/Models/Config/Role.php:106]: I don't understand these methods. Did you want to do something like this?

```suggestion
  public function getCreatedByUserAttribute()
  {
    return \App\v1\Models\Common::getUserAttributes($this->created_by);
  }

  public function getUpdatedByUserAttribute()
  {
    return \App\v1\Models\Common::getUserAttributes($this->updated_by);
  }

  public function getDeletedByUserAttribute()
  {
    return \App\v1\Models\Common::getUserAttributes($this->deleted_by);
  }
```

File: src/v1/Controllers/Item.php [hunk 5]
```diff
@@ -165,6 +165,8 @@ public function getAll(Request $request, Response $response, $args): Response
     $token = (object)$request->getAttribute('token');
     $organizations = \App\v1\Common::getOrganizationsIds($token);
     $parentsOrganizations = \App\v1\Common::getParentsOrganizationsIds($token);
+    // check permissions
+    \App\v1\Controllers\Config\Role::checkPermissionToData('view', $args['typeid']);
```
Comment (marien-probesys) [inline src/v1/Controllers/Item.php]: As a developer discovering the code, I don't expect the methods that check the permissions to be declared in controllers. In fact, all these methods don't belong to controllers but a class dedicated to permissions (or maybe several):

- Role::generatePermission
- Role::getStructureViewIds
- Role::checkPermissionToStructure
- Role::checkPermissionToData
- Permissiondataproperty::getPropertiesCanView
- Permissionstructurecustom::getPropertiesCanView

(btw, the two last methods do the same thing)

Something like `App\v1\Permission` or `App\v1\Access`?

File: src/v1/Models/Config/Permissionstructurecustom.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+  public function getSoftdeleteAttribute($value)
+  {
+    return boolval($value);
+  }
+
+  public function getDeleteAttribute($value)
+  {
+    return boolval($value);
+  }
```
Comment (marien-probesys) [inline src/v1/Models/Config/Permissionstructurecustom.php:81]: Same as for the `Role` model, all these methods do the same thing and I don't understand how they work

**Review by marien-probesys — COMMENTED** [review]
If I try to summarize:

- permissions to manage **items** are handled by the **Permissiondata** model
- permissions to manage **properties** are handled by the **Permissiondataproperty** model
- and permissions to manage **types/structure** are handled by the **Permissionstructure** model (and Permissionstructurecustom? but I didn't get why we need it)
- the permissions are attached to a role, which is given to a user (an `Item`)
- the role also has global `permissionstructure` and `permissiondata` to grant or remove permissions all at once

What bothers me is that we don't need all this complexity for the moment. The Role model is globally good to me, and Permissiondata as well. In fact, they are enough to verify if a user can list tickets, or if he can create users for instance. The rest of the code is dead code because we'll not use it before a long time.

We could easily decide that `Role.permissionstructure` is enough to manage permissions on the structure (with `permissionstructure=grant` set only for the admin user), and I don't think we need specific permissions for properties at the moment.

What do you think?

**Comment by ddurieux** [comment]
You can have in role permissiondata and permissionstructure to grant to allow all rights. It can be enough, but you can too have more specifics rights on items / config...

File: src/v1/Models/Item.php (hunk 1 above)
Comment (ddurieux) [reply src/v1/Models/Item.php:221]: Like we discuss, the next step is to associate the role with the organization, so will be able to have multiple roles

File: src/v1/Controllers/Token.php (hunk 2 above)
Comment (ddurieux) [reply src/v1/Controllers/Token.php]: Fixed

File: src/v1/Models/Item.php (hunk 3 above)
Comment (ddurieux) [reply src/v1/Models/Item.php]: fixed

File: src/v1/Models/Config/Role.php (hunk 4 above)
Comment (ddurieux) [reply src/v1/Models/Config/Role.php:106]: It's the system of the eloquent model, the function name is a norme base on field name

File: src/v1/Controllers/Item.php (hunk 5 above)
Comment (ddurieux) [reply src/v1/Controllers/Item.php]: Good idea, I put into App\v1\Permission

File: src/v1/Models/Config/Permissionstructurecustom.php (hunk 6 above)
Comment (ddurieux) [reply src/v1/Models/Config/Permissionstructurecustom.php:81]: it convert the value of the DB into value to the controller (here convert 0 or 1 to a boolean)

**Pull request merged**